[SYCL] Fix crash caused by functor without call operator as Kernel

srividya-sundaram opened the pull request:
When a named function object without a call operator defined is used as a kernel, the compiler causes a `crash`.
With the changes in this PR, a compiler `error` is thrown when a lambda or function object without a call operator defined is used as a kernel.

smanna12 commented on clang/test/SemaSYCL/undefined-functor.cpp:
```diff
@@ -0,0 +1,18 @@
+// RUN: %clang_cc1 -fsycl-is-device -internal-isystem %S/Inputs -sycl-std=2020 -verify -fsyntax-only %s
+
```
Please add a comment about what test is doing?

premanandrao commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -2412,6 +2412,19 @@ static bool isESIMDKernelType(const CXXRecordDecl *KernelObjType) {
+    CallOperator = getOperatorParens(KernelObjType);
+  return (CallOperator != nullptr);
+}
+
```
Looks good to me.

Around line 2512, there is identical logic used for getting the call operator.   Can you refactor this function (perhaps by having it return the CallOperator) so it can be called there as well?

premanandrao commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -976,6 +976,27 @@ static QualType GetSYCLKernelObjectType(const FunctionDecl *KernelCaller) {
+
+// Fetch the associated call operator of the kernel object
+// (of either the lambda or the function object).
+CXXMethodDecl *IsCallOperatorDefined(const CXXRecordDecl *KernelObjType) {
```
Perhaps change the name to GetCallOperatorIfDefined or something?   It is no longer a bool function - so IsBlah does not seem right.

premanandrao commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -3454,10 +3464,11 @@ void Sema::CheckSYCLKernelCall(FunctionDecl *KernelFunc, SourceRange CallLoc,
     return;
 
-  const CXXRecordDecl *KernelObj =
+  CXXRecordDecl *KernelObj =
```
Is the loss of 'const' here inadvertent?

srividya-sundaram replied:
Dropped it intentionally.
Looking at the documentation of  [const](https://en.cppreference.com/w/c/language/const)  and the fact that  [GetSYCLKernelObjectType](https://github.com/intel/llvm/blob/3eecf940a3c136d557c9fad705961912cdc82102/clang/lib/Sema/SemaSYCL.cpp#L967) doesn't return a `const-qualified` type, I didn't see the need to have this `const-qualified`.
Besides, I see `const` qualifier [applied](https://github.com/intel/llvm/blob/3eecf940a3c136d557c9fad705961912cdc82102/clang/lib/Sema/SemaSYCL.cpp#L3631) in some cases and [left](https://github.com/intel/llvm/blob/3eecf940a3c136d557c9fad705961912cdc82102/clang/lib/Sema/SemaSYCL.cpp#L2511) other in others. 
If there is a rationale behind adding the `const` qualifier, I'd like to know.
If not, I can remove them at other places too.

elizabethandrews replied:
I think declaring something as `const` allows for compiler optimizations sometimes, but I'm not sure if those happen here. One advantage of declaring the KernelObj as const is it avoids us from accidentally modifying it. @premanandrao thoughts?

aelovikov-intel replied:
I don't believe `const` can help optimizations: https://www.youtube.com/watch?v=FnGCDLhaxKU&t=5730s .

elizabethandrews replied:
Very interesting. Thanks for sharing @aelovikov-intel

srividya-sundaram replied:
Not sure if this is of any relevance to why the KernelObject was `const` qualified, but the SYCL [spec](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:interfaces.kernels.as.function-objects) mentions that : `The operator() member function must be const-qualified` , when kernels are defined as functors.

premanandrao replied:
> Dropped it intentionally. Looking at the documentation of [const](https://en.cppreference.com/w/c/language/const) and the fact that [GetSYCLKernelObjectType](https://github.com/intel/llvm/blob/3eecf940a3c136d557c9fad705961912cdc82102/clang/lib/Sema/SemaSYCL.cpp#L967) doesn't return a `const-qualified` type, I didn't see the need to have this `const-qualified`. Besides, I see `const` qualifier [applied](https://github.com/intel/llvm/blob/3eecf940a3c136d557c9fad705961912cdc82102/clang/lib/Sema/SemaSYCL.cpp#L3631) in some cases and [left](https://github.com/intel/llvm/blob/3eecf940a3c136d557c9fad705961912cdc82102/clang/lib/Sema/SemaSYCL.cpp#L2511) out in other cases. If there is a rationale behind adding the `const` qualifier, I'd like to know. If not, I can remove them at other places too.

I would like the const re-applied in this case.  The rationale is for both visual affirmation that we are not modifying the object pointed to as well as for potential optimization.   (There are arguments that it does not really help with optimization; but it is not clear to me if that is a statement of the current state of the optimizer or a stronger statement that it can never be optimized.)

I would err on the other side.   I would apply const qualifiers to other places that do not have them but should; rather than remove it from all places.

premanandrao replied:
> I don't believe `const` can help optimizations: https://www.youtube.com/watch?v=FnGCDLhaxKU&t=5730s .

Thanks for that link @aelovikov-intel.   I remember seeing it a few years back, it is a good talk.  :-)

premanandrao replied:
> Not sure if this is of any relevance to why the KernelObject was `const` qualified, but the SYCL [spec](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:interfaces.kernels.as.function-objects) mentions that : `The operator() member function must be const-qualified` , when kernels are defined as functors.

That wouldn't affect the internal compiler structure.   That const-ness is at the source level.

premanandrao commented on clang/test/SemaSYCL/undefined-functor.cpp:
```diff
@@ -0,0 +1,19 @@
+using namespace sycl;
+queue q;
+
+struct FunctorWithoutCallOperator; // expected-note {{forward declaration of 'FunctorWithoutCallOperator'}}
```
Can you add another variant case where the struct is defined but without the call operator?

premanandrao commented on clang/test/SemaSYCL/undefined-functor.cpp:
```diff
@@ -0,0 +1,19 @@
+struct FunctorWithoutCallOperator; // expected-note {{forward declaration of 'FunctorWithoutCallOperator'}}
+
+int main() {
+    // expected-error@#KernelSingleTask {{kernel parameter must be a lambda or function object}}
```
Nit: comment indentation

premanandrao commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -3454,10 +3464,11 @@ void Sema::CheckSYCLKernelCall(FunctionDecl *KernelFunc, SourceRange CallLoc,
       GetSYCLKernelObjectType(KernelFunc)->getAsCXXRecordDecl();
 
-  if (!KernelObj) {
+  CXXMethodDecl *CallOperator = IsCallOperatorDefined(KernelObj);
+  if (!CallOperator) {
```
```suggestion
  if (!IsCallOperatorDefined(KernelObj)) {
```
(and change the name to whatever we call it above).

elizabethandrews commented on clang/test/SemaSYCL/undefined-functor.cpp:
```diff
@@ -0,0 +1,30 @@
+  q.submit([&](sycl::handler &cgh) {
+    // expected-error@#KernelSingleTask {{kernel parameter must be a lambda or function object}}
+    // expected-note@+1 {{in instantiation of function template specialization}}
+    cgh.single_task(StructDefined{});
```
So the error is triggered for declarations but not definitions? Is the error valid then? Should we be diagnosing incomplete types instead?

srividya-sundaram replied:
We are triggering the following error `kernel parameter must be a lambda or function object`, whenever a Functor (whether just declared or defined)  does not have the call operator defined. The key is "call operator" being defined in the Functor.

This error - `invalid use of incomplete type 'FunctorWithoutCallOperator'` is an existing C++ error (unrelated to SYCL) that gets triggered when we have incomplete types used in the code - `struct FunctorWithoutCallOperator` in our case.

Since `struct StructDefined` is completely defined, we don't see the error for that case.

Hope this answered your question.

premanandrao replied:
What if the call operator is simply declared, but not defined in the function object?

elizabethandrews replied:
>> Should we be diagnosing incomplete types instead?
>>> This error - invalid use of incomplete type 'FunctorWithoutCallOperator' is an existing C++ error (unrelated to SYCL) that gets triggered when we have incomplete types used in the code

For some reason I thought you implemented a new diagnostic talking about missing call operators. The diagnostic is for incomplete types like I thought it should be. Sorry for the confusion.

I do have a follow-up question though. It's not clear to me why the compiler crashes without your patch. Shouldn't the diagnostic for incomplete types get trigger even without your patch? It looks like your patch is just adding a second diagnostic to same site.

elizabethandrews replied:
Actually my bad. The diagnostic is at the single_task definition, not the instantiation. My question about the crash remains though. Can you explain why it crashes?

srividya-sundaram replied:
> It's not clear to me why the compiler crashes without your patch. Shouldn't the diagnostic for incomplete types get triggered even without your patch? 

Prior to my patch, the compiler would crash(from an assert failure) **AND** emit the error -  `invalid use of incomplete type 'foo'` for the following code:

```
#include "CL/sycl.hpp"
struct foo;
int main() {
  sycl::queue{}.single_task(foo{});
}
```

With my patch, I am replacing the crash with the error - `kernel parameter must be a lambda or function object`, which is why we now get 2 errors for the above code.

The error - `kernel parameter must be a lambda or function object`, was already getting triggered for invalid kernel objects. In this patch, I additionally check if the named function object has a function call operator declared or defined in it and if not, I trigger the same error.

Does this help clarify your questions?

srividya-sundaram replied:
> What if the call operator is simply declared, but not defined in the function object?

The compiler does not trigger an error if the call operator is simply declared. 
I have updated the test with two additional test cases - Functor with just the call operator declaration and one with definition.

srividya-sundaram commented:
ping @intel/dpcpp-cfe-reviewers

premanandrao commented on clang/test/SemaSYCL/undefined-functor.cpp:
```diff
@@ -0,0 +1,50 @@
+  int x;
+};
+
+class FunctorWithCallOpDeclared {
```
Nit: the names should be the other way round.   This one is 'FunctorWithCallOpDefined'.

srividya-sundaram replied:
Duh! Good catch.

premanandrao commented on clang/test/SemaSYCL/undefined-functor.cpp:
```diff
@@ -0,0 +1,50 @@
+  void operator()() const {}
+};
+
+class FunctorWithCallOpDefined {
```
Swap name with the previous one.

premanandrao reviewed (commented):
Can you make the changes in c7798556b624e901e297a587619ea2b5ee03d489 a separate NFC PR?  Not needed as part of this one.

srividya-sundaram commented:
Failures in `SYCL / Linux / OCL GEN9 LLVM Test Suite`  and `SYCL / Linux / OCL x64 LLVM Test Suite `are unrelated to this patch.

premanandrao commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -976,6 +976,28 @@ static QualType GetSYCLKernelObjectType(const FunctionDecl *KernelCaller) {
+
+// Fetch the associated call operator of the kernel object
+// (of either the lambda or the function object).
+CXXMethodDecl *
```
Really sorry for not catching this earlier.
```suggestion
static CXXMethodDecl *
```

srividya-sundaram replied:
No worries. Thanks for the reviews!

premanandrao commented on clang/test/SemaSYCL/undefined-functor.cpp:
```diff
@@ -0,0 +1,50 @@
+class FunctorWithCallOpDeclared {
+  int x;
+  public:
+  void operator()() const {};
```
```suggestion
  void operator()() const;
```

srividya-sundaram commented:
@intel/llvm-gatekeepers Hi, this is ready to be merged. Thanks!

steffenlarsen commented:
@srividya-sundaram - CI seems to fail in SemaSYCL/undefined-functor.cpp.

srividya-sundaram commented:
@steffenlarsen Thanks for pointing out. I have fixed the failures.

I have removed the test case causing failure for the below reasons:
We currently diagnose undefined functions called on device from either kernels or SYCL device functions unless they are marked with the SYCL_EXTERNAL attribute. We could either add the SYCL_EXTERNAL attribute or expect the error to fix the test.
But this seems like a tangent to what we want to test in this patch.
Also, lambda expressions(closure types) and functors, which are used to define SYCL kernels, have the call operator defined in them. Hence dropping the test case with just the call operator declared seemed appropriate.

premanandrao commented:
> I have removed the test case causing failure

What was the nature of the failure?

srividya-sundaram commented:
> > I have removed the test case causing failure
> 
> What was the nature of the failure?

The test was throwing this error for the undefined functor  case:
`SYCL kernel cannot call an undefined function without SYCL_EXTERNAL attribute error: 'note' diagnostics seen but not expected`

srividya-sundaram commented:
ping @intel/llvm-gatekeepers

pvchupin commented:
OCL x64 testing failed on drivers download attempt. Looks like some sporadic network issue.
